lib/SILOptimizer/Mandatory/TFUtilities.h

```diff
@@ -470,10 +475,21 @@ struct GraphGlobalConfiguration {
     static std::pair<StringRef, SILTensorOpInfo::OperandClass>
     decodeAttributeName(Identifier name);
 
-  private:
+    /// Get an int-typed attribute at `attrIdx`, which must have `attrName`.
+    int64_t getIntAttr(unsigned attrIdx, StringRef attrName) const;
+
+    /// Get a string-typed attribute at `attrIdx`, which must have `attrName`.
+    std::string getStringAttr(unsigned attrIdx, StringRef attrName) const;
```

Comment:
I still don't understand why you need to have these helper functions.  I'd prefer to not have them in the GraphOperationInfo class if they are needed.  Can they become static helpers closer to the code that is using them?

Reply:
(posted a reply earlier on but it might get missed from the github UI)

A few reasons for having these APIs:

- GraphOperationInst::getAttribute(name) (or the renamed version) uses a loop to find such an attr name -- it could be changed to a hash table in the future though.
- For synthesized attrs (like tensor ids from sends / recvs and device partitioning), the code would be more robust (but also more rigid admittedly), if the caller gets to assert that not only an attr is present, but it must be in a certain position / idx.

I can move to the call-site if you prefer.

Reply:
There are call-sites from TFDevicePartition.cpp and TFLowerGraph.cpp, so TFUtilities.cpp is a reasonable place to place this helper method. Let me know what you think.

Reply:
Ok, we can revisit this in future patches.